### PR TITLE
STARCH-972 Runtime Environment and Log Fine-Tuning

### PR DESCRIPTION
**Pull request by mwfyffeiu**

### CODE REVIEW

File: client/src/main/java/iu/client/JsonSerializer.java [hunk 1]
```diff
@@ -96,11 +96,13 @@ static String formatPropertyName(String propertyName, IuJsonPropertyNameFormat p
...
 	 * @param propertyNameFormat property name format
 	 * @param adapt              adapter function
-	 * @return {@link JsonObject}
+	 * @return {@link JsonObject}; null if value is null
 	 */
 	@SuppressWarnings({ "unchecked", "rawtypes" })
 	public static <T> JsonObject serialize(Class<T> type, T value, IuJsonPropertyNameFormat propertyNameFormat,
 			Function<Type, IuJsonAdapter<?>> adapt) {
+		if (value == null)
+			return null;
```
Comment (james-anderson-iu) [inline client/src/main/java/iu/client/JsonSerializer.java]: **🟡 IMPORTANT — Returning Java null for a null bean makes IuJsonAdapter.from(...).toJson(null) diverge from every other adapter, silently dropping map entries**

JsonSerializer.serialize is the toJson body of the public adapter returned by IuJsonAdapter.from(Class, IuJsonPropertyNameFormat, Function) (IuJsonAdapter.java:227). Every other adapter in this module maps a null Java value to JsonValue.NULL: BasicJsonAdapter.toJson lines 100-101, and JsonObjectAdapter.toJson lines 82-83 in this same chunk. This new early return makes one adapter return Java null instead. Downstream, IuJson.add skips a null value outright (IuJson.java:289), so a bean-typed value that is null inside a Map serialized by JsonObjectAdapter (line 87 of that file, also changed here) disappears from the object rather than serialising as null. Before this change, line 107's value.getClass() raised NullPointerException on the same input - a loud failure has been replaced with silent omission.

**Impact:** A remote peer receives a JSON object in which a key is absent rather than null. Absent and null are not interchangeable for most consumers - a partial-update payload in particular reads an absent key as 'leave unchanged' and an explicit null as 'clear'. Nothing on either side reports the difference.
**Conditions:** Requires the bean adapter to be used as a VALUE or ITEM adapter - a map value, a collection or array element, an Optional payload, or a remote method parameter. A bean's own properties are unaffected: JsonSerializer.serialize adds each property through IuJson.add(builder, name, () -> propertyValue, adapter) (IuJson.java:283-291), whose guard tests the Java value for null before the adapter is consulted, so the documented 'properties with a null value are skipped' contract holds independently of this defect and of its fix.
**Fix:** serialize's JsonObject return type cannot express JsonValue.NULL, so fix it at the adapter boundary: change IuJsonAdapter.java:227 to v -> v == null ? JsonValue.NULL : JsonSerializer.serialize(type, v, propertyNameFormat, valueAdapter), matching the null handling its own fromJson lambda already has on line 226. Blast radius traced across iu-java-util, esas-java-util, ess-thirdparty and esas-sisjee: (1) bean properties are unchanged, so MetadataServlet.java:88, IuVaultClient.java:37,45 and IuConfig.java:270-278 all emit byte-identical output; (2) map values change from absent to explicit null, via JsonObjectAdapter.java:87 -> IuJson.add(JsonObjectBuilder,String,Object) at IuJson.java:240, whose guard tests the returned JsonValue rather than the Java value - this is the only wire-format change; (3) collection, array and Optional elements change from a NullPointerException to null, because JsonArrayAdapter.java:98 passes the adapter result straight to JsonArrayBuilder.add(JsonValue), which rejects null per the Jakarta JSON specification, and CollectionAdapter extends JsonArrayAdapter. No caller in any of the four repositories branches on a null toJson result (greps for 'toJson(...) == null', '!= null' and requireNonNull over main sources returned nothing), so nothing depends on the current return. Also update the from(Class, ...) javadoc to state that toJson(null) returns JsonValue.NULL; JsonSerializer.serialize's own '@return JsonObject; null if value is null' stays accurate because serialize is unchanged.
**Verify:** Add three tests: a Map<String, Bean> with a null value, asserting the key is present as JsonValue.NULL rather than absent; a List<Bean> with a null element, which throws NullPointerException today and must serialize as null after the fix; and a remote invocation with a null bean-typed argument through RemoteInvocationHandler.serialize, which is the reachable production path that currently fails. Guard MetadataServlet.java:88 against a null metadata if belt-and-braces is wanted: today a null would raise NullPointerException on .toString(), after the fix it would serve the literal string 'null' as the discovery document.
**Verified** against IuJsonAdapter.from delegates toJson to JsonSerializer.serialize, IuJson.add omits a null value outright, the pre-change line 107 raised NullPointerException on the same input, second-pass blast-radius trace: RemoteInvocationHandler.java:529 passes the adapter result straight to JsonArrayBuilder.add, and adapt (lines 584-590) returns from(Class, ...) for any non-platform parameter type, so a remote call with a null bean-typed argument throws NullPointerException on the client today - a reachable production failure, not only a silent omission, bean properties are guarded on the Java value at IuJson.java:283-291, so no bean's own serialized output changes.
Confidence: high

**Review by james-anderson-iu — COMMENTED** [review]
## Automated Java review — linear pipeline

18 files reviewed across `logging`, `base`, `el` and `client`; 5 lens passes (security, correctness, performance, craftsmanship, challenge) plus a second verification stage.

**Verdict: Needs changes** — 0 CRITICAL, 4 IMPORTANT, 10 SUGGESTION.
Verification: 5 confirmed, 2 narrowed, 0 retracted, 0 unresolved. 21 candidate issues were investigated and ruled out.

Findings anchored to specific lines are inline. The rest follow, either because the anchor is not on a changed line or because the finding covers more than one location.

---

**🟡 IMPORTANT — New runtime-environment helpers document an environment variable but resolve a system property first, understating the trust surface of the switches they gate**

flag, bound and longBound all delegate to envOptional, which is System.getProperty(name, System.getenv(env)) - the system property wins and the environment variable is only the fallback. The pre-existing accessors say so in their own javadoc ('Checks for the presence of a system property, then defaults to an environment variable', lines 51-52 and 85-86). Every one of the three new helpers instead documents only '@param name environment variable name' (lines 131, 148, 169, 184, 197), and flag's body text frames these as switches that 'turn off a protection or trust a header'. An operator who reads the accessor that is actually named in the code and hardens the process environment accordingly will not learn from it that -Dname=true on the command line, or any code on the class path calling System.setProperty, reaches the same switch - and unlike the environment, a system property is mutable for the life of the JVM. Beyond the trust-surface point, the prose itself does not belong to this class: flag says 'Every one of these turns off a protection or trusts a header' - there is no 'these' in edu.iu, and nothing in this module trusts headers; bound says 'Shared by every numeric setting the container accepts' - edu.iu has no container, and the only numeric settings in this repository that could use bound (IuLogHandler.java:127-130, 170-173) do not. The wording reads as if lifted from the servlet-container class that previously owned these helpers, which the consumer checkout corroborates. The two bound overloads then carry byte-identical javadoc paragraphs, and neither documents that maxValue is enforced.

**Impact:** A protection switch believed to be controlled by the deployment environment can be set from the JVM command line or from in-process code. The observed consumer uses flag for an authentication bypass (esas-java-util IuTomcatServerTest.java:728, with the production read at IuTomcatRuntimeEnvironment.java:370), so the switch whose documentation is least accurate is also the highest-value one. Separately, a reader of edu.iu.IuRuntimeEnvironment is told about a container and a set of sibling settings that do not exist here, and is not told the one thing the @throws clause needed to say. Documentation that describes the wrong system is worse than none, because it is trusted.
**Conditions:** Applies wherever an operator relies on the new helpers' javadoc rather than on envOptional's. The system-property-first precedence is pre-existing and unchanged; only the documentation of the new entry points is wrong.
**Fix:** Reword @param on flag, bound (x2) and longBound (x2) to match envOptional - 'property name; checked as a system property first, then as the corresponding environment variable' - and state that a system property overrides the environment. If a security switch must not be settable via -D, read it through System.getenv directly rather than through envOptional. Additionally, rewrite the three javadoc blocks in terms of what this class does rather than the container they came from, differentiate the two bound overloads, and add maxValue to the @throws on both.
**Verify:** Add a test asserting that a system property set after JVM start is observed by flag(name) even when the corresponding environment variable is absent, so the precedence is pinned rather than only documented.
**Verified** against envOptional resolves System.getProperty(name, System.getenv(env)) - the system property wins, the pre-existing accessors document that precedence and the three new helpers do not, the consumer's bypass switch is only a constant name passed to IuRuntimeEnvironment.flag, the switch is read from production code at IuTomcatRuntimeEnvironment.java:370, not only from test mocks, and gates whether the access-control filter is installed at all (lines 903-912).
**Also at:** `base/src/main/java/edu/iu/IuRuntimeEnvironment.java:155`
*Found by the security and craftsmanship passes.*
Confidence: high

---

**🟢 SUGGESTION — Bootstrap.getEnvironment reads the ENVIRONMENT WeakHashMap without the lock every writer holds**

ENVIRONMENT is a plain WeakHashMap (line 70). initializeContext correctly serialises its check-then-put under synchronized (ENVIRONMENT) at lines 186-193, which establishes that the author considers the map to need a lock. getEnvironment reads the same map with no lock and no memory barrier. A HashMap-family read concurrent with a write can observe a partially rehashed table: at best a stale or missing entry, at worst a get() that never terminates. This is on the hot path - every ProcessState construction calls it (ProcessLogger.java:143), so every follow does.

**Impact:** A thread serving traffic can hang inside getEnvironment, or silently fall back to the PLATFORM defaults and stamp log events with the wrong application, environment and endpoint - which is also the key IuLogHandler uses to choose a log file (IuLogHandler.java:244-250), so records land in the wrong file.
**Conditions:** Unreachable in every observed deployment, because no consumer calls initializeContext and the map is therefore never written. Becomes live only for an application that calls the published IuLogContext.initializeContext API: then either a write for one context class loader overlaps a follow on another thread, or - with no writer at all - concurrent getEnvironment reads expunge a stale entry after that application is undeployed. Pre-existing rather than introduced here.
**Fix:** Replace WeakHashMap with Collections.synchronizedMap(new WeakHashMap<>()) and route both accesses through it, or hold synchronized (ENVIRONMENT) around the get in getEnvironment.
**Verify:** Add a test that hammers getEnvironment from several threads while initializeContext registers new context class loaders, asserting every read returns either PLATFORM or a fully constructed LogEnvironmentImpl.
**Narrowed** from IMPORTANT against no approved consumer repository references initializeContext at all - the ENVIRONMENT map is never written in any observed deployment, so getEnvironment always returns PLATFORM and there is no writer to race, the tomcat server calls IuLogContext.initialize() at IuTomcatBootstrap.java:587, the platform-level entry point, not the per-context initializeContext, initializeContext is idempotent-by-rejection: it throws IllegalStateException when an entry already exists for the context class loader (Bootstrap.java:187-189) and when the context loader is the platform or system loader (lines 177-180), so it cannot be called twice for one loader, residual mechanism found in the second pass: WeakHashMap.get itself mutates, because getTable calls expungeStaleEntries, so once any entry exists and its class loader is collected, concurrent getEnvironment calls alone can structurally modify the map with no concurrent writer at all.
Confidence: high

---

**🟢 SUGGESTION — Two redundant ACTIVE_PROCESS_STATE.remove() calls sit in the exact block a reader must reason about for thread-local lifetime**

Line 357 removes the thread-local and line 358 immediately sets it - the remove has no effect, since set overwrites unconditionally. Line 369 removes it again on the way to the return, and the finally at 373-378 then runs and unconditionally re-establishes the correct value, so that call has no effect either. Both are no-ops, and both are in the fifteen lines that a reviewer or a future maintainer reads most carefully, because this is where a leaked frame would come from. A no-op remove next to a real one invites the reading that the removes are load-bearing and the finally is a backstop, which is the reverse of the truth.

**Impact:** Costs a reader time in the one place in this file where the cost is highest, and makes the next change to the lifecycle harder to get right.
**Conditions:** Applies as written; no runtime condition.
**Fix:** Delete the remove at 357 and the remove at 369; the try/finally at 354-378 already expresses the whole contract.
**Verify:** Existing ProcessLoggerTest cases continue to pass, including the fork/join assertions that the process is not left bound to a thread.
Confidence: high

---

**🟢 SUGGESTION — IuLogHandler keeps its own copies of the helpers this change added to IuRuntimeEnvironment, in a file the same change edits**

This private helper is IuRuntimeEnvironment.envOptional(name, convert) with a default substituted for null - the same shape as the envOptional(String, Function) overload the change just added two modules down. Lines 127-130 of the same file hand-roll a bound as Long.parseLong(Objects.requireNonNullElse(envOptional("iu.logging.file.maxSize"), "10485760")) and Integer.parseInt(... "10"), which is what the new longBound and bound exist to replace, minus their validation: as written, iu.logging.file.nLimit = 0 or a negative maxSize is accepted and fails later at the file publisher instead of at the point of configuration. Adding shared helpers and leaving the duplicate in a file the same commit touches is how the duplicate survives the next refactor.

**Impact:** Two implementations of the same configuration read will drift, and the logging module's numeric settings keep reporting bad values differently from every caller that adopts the new helpers - which is the difference the new javadoc claims to have removed.
**Conditions:** Applies as written; no runtime condition.
**Fix:** Replace the private env helper with IuRuntimeEnvironment.envOptional(name, convert) plus Objects.requireNonNullElse, and read maxSize and nLimit through longBound and bound so an out-of-range value is rejected where it is configured.
**Verify:** Add a test asserting iu.logging.file.nLimit = 0 is rejected with a message naming the property.
Confidence: high

File: logging/impl/src/main/java/iu/logging/internal/ProcessLogger.java [hunk 2]
```diff
@@ -405,7 +494,10 @@ public static void trace(Supplier<String> messageSupplier) {
 		if (state == null)
 			return;
 
-		state.children.offer(new TracedMessage(message, state.depth));
+		final var joinId = ACTIVE_JOIN_ID.get();
+		state.children.offer(new TracedMessage( //
```
Comment (james-anderson-iu) [inline logging/impl/src/main/java/iu/logging/internal/ProcessLogger.java]: **🟡 IMPORTANT — The process trace grows without bound for the life of a followed process, and every WARNING copies the whole of it**

children has no cap and no eviction, and IuLogHandler.publish feeds it every record the handler publishes (IuLogHandler.java:357-361), so its size tracks the total log volume of the process rather than anything the caller chose to trace. The state is reachable from the thread-local for the whole follow scope. Independently, IuLogEvent captures processLog = ProcessLogger.export() for every record at WARNING or above (IuLogEvent.java:116-117) and appends it to the formatted output (lines 200-203), so with n accumulated entries and w warnings the process costs O(n*w) time and retains O(n*w) characters across the queued events - the events themselves are capped by count (maxEvents, default 100000) but not by size. The new fork/join surface raises n's growth rate directly: several worker threads may now offer into one process's children concurrently (ProcessLogger.join lines 428-441), where previously only the following thread could.

**Impact:** For a follow scope longer than a single short request - a batch job, a scheduled task, a streaming or long-poll handler, or a fan-out that joins many workers - heap held by the trace and by the warning events that copied it grows with log volume, and each warning writes an ever-larger block into the error and info log files. At a few thousand accumulated entries a single warning writes hundreds of kilobytes. Under an error burst on the observed per-request consumer, retained heap grows with the error rate because the event buffer evicts by count rather than by bytes.
**Conditions:** Requires a followed process that is long-lived or high-volume relative to the number of WARNING+ records it produces. The observed consumer scopes a process to one HTTP request (esas-java-util IuTomcatRequestFilter.java:831), which bounds n to that request's log records, so the O(n*w) half does not bite there today; the retention half does, and the unbounded-scope half bites the first caller that wraps a non-request unit of work.
**Fix:** Cap children - either a maximum entry count with a dropped-message marker, or a maximum retained character budget - and reuse a memoised trace rendering for repeated export() calls within a process instead of rebuilding it per warning.
**Verify:** Add a test that traces well past the cap inside one follow scope and asserts both that export() stays within the budget and that a truncation marker appears; benchmark export() cost against trace size to confirm it no longer scales with the number of warnings.
**Verified** against the event buffer is capped by maxEvents count, not by retained bytes, each WARNING+ record captures and retains its own full ProcessLogger.export() copy, the observed consumer scopes a process to one HTTP request.
Confidence: high

File: logging/impl/src/test/java/iu/logging/internal/ProcessLoggerTest.java [hunk 3]
```diff
@@ -55,6 +58,17 @@
 @SuppressWarnings("javadoc")
 public class ProcessLoggerTest {
 
+	@BeforeEach
```
Comment (james-anderson-iu) [inline logging/impl/src/test/java/iu/logging/internal/ProcessLoggerTest.java]: **🟡 IMPORTANT — The new ProcessLogger tests pin only the success paths of the thread-local lifecycle they were written for**

174 added lines cover fork, nested join, join on the forking thread, and the terminal-reference export fix - all on paths where nothing throws. The parts of follow and join that exist purely for the abrupt-completion case, the finally blocks at ProcessLogger.java:373-378 and 443-452, are never executed by any test in the file. Nor is the ClassCastException path in join (line 429) for a handle that did not come from fork. Separately, the @BeforeEach resets only REQUEST_ID; ACTIVE_PROCESS_STATE and ACTIVE_JOIN_ID are left as the previous test left them, so a genuine leak of either would silently contaminate the next test rather than fail it - the file's own assertions such as 'process left bound to the worker thread' (line 266) show the author intends to catch exactly that class of bug.

**Impact:** The regression that this change is most exposed to - a frame left on a pooled thread after a failure, which is a cross-request context bleed in a shared library - would ship green.
**Conditions:** Applies as written; no runtime condition.
**Fix:** Add cases for follow with a throwing supplier and join with a throwing task, each asserting fork() is null (or the prior handle) afterwards; assert in @BeforeEach that ProcessLogger.fork() and the join ID are already clear, so a leak from an earlier test fails at its own boundary.
**Verify:** The added tests fail if the finally blocks are removed and pass with them present.
**Verified** against no throwing follow or join exists across all three test classes in the change, BootstrapTest mocks ProcessLogger out entirely, the IT's throwing branch is conditional on a worker failure that does not occur on a passing run, second-pass check against the reported 100% coverage: ProcessLoggerTest contains no assertThrows and only one throw, at line 152, which rethrows a worker thread's uncaught exception in the test harness rather than inside a follow or join body, a finally block runs on normal return, so the blocks at ProcessLogger.java:373-378 and 443-452 reach full line coverage without any exception ever unwinding through them, and the if (restore == null) branch inside gets both outcomes from the top-level and nested follow tests - so 100% line and branch coverage is exactly what a suite with no abrupt-completion case produces.
Confidence: high

File: logging/impl/src/main/java/iu/logging/internal/ProcessLogger.java [hunk 4]
```diff
@@ -346,18 +355,18 @@ public static <T> T follow(LogContext context, String header, UnsafeSupplier<T>
...
 			final var rv = supplier.get();
 
 			state.end();
 
-			ACTIVE_PROCESS_STATE.remove();
 			if (restore == null)
 				LOG.info(() -> "complete " + state.requestId + ": " + header + System.lineSeparator() + state);
 			else
 				LOG.info(() -> "end " + state.requestId + ": " + header);
+			ACTIVE_PROCESS_STATE.remove();
```
Comment (james-anderson-iu) [inline logging/impl/src/main/java/iu/logging/internal/ProcessLogger.java]: **🟢 SUGGESTION — The process trace is emitted only when the followed task succeeds, so a failed process is the one case that produces no trace**

state.end() at line 363 and the 'complete'/'end' LOG.info at 365-368 sit inside the try, after supplier.get(). When the body throws, none of them run, so the accumulated trace is never logged and never reaches a log file. The finally at 373-378 then restores the thread-local before the exception leaves follow, so a caller that logs the failure after follow returns has already lost the binding: IuLogEvent captures ProcessLogger.export() for WARNING and above (IuLogEvent.java:116-117), and by then export() sees the enclosing process, or null. For a library whose stated purpose is post-mortem process tracing, the trace is dropped in precisely the situation it exists for.

**Impact:** A caller that lets the followed body propagate produces no process trace at all, while every successful one does. Diagnosis of the failure falls back to whatever the caller logged on its own. The observed consumer absorbs Throwable inside the boundary, so it is unaffected today.
**Conditions:** Requires the UnsafeSupplier passed to follow to complete abruptly. IuLogContext.follow declares throws Throwable and IuLoggingBootstrapIT.testInit (lines 100-116) exercises a body that can throw, so the path is real; the dominant observed caller does not reach it.
**Fix:** Move state.end() and the trace emission into a finally (or a catch that rethrows), so the process is closed out and logged on both paths - logging at the context's level, or at WARNING, when the body threw.
**Verify:** Add a ProcessLoggerTest case whose follow body throws, asserting a 'complete N' record is still published with the accumulated trace and that ProcessLogger.fork() is null afterwards.
**Narrowed** from IMPORTANT against the outer consumer call site catches Throwable inside the follow body at IuTomcatRequestFilter.java:858-861 and logs the failure there, IuLogEvent captures ProcessLogger.export() for WARNING and above, so the trace is captured inside the boundary, the second call site, IuTomcatSecurityContext.java:117, does NOT guard its body - the filter-chain call propagates rather than being caught - but it runs nested inside the request filter's process, since it reads the REQUEST_ID that filter set, ProcessState registers itself in the enclosing process's children at construction (ProcessLogger.java:132-137), before the body runs, and export walks children recursively (lines 183-207), so a nested process's traced messages survive in the parent trace even when the nested body throws, ess-thirdparty has no follow call sites at all.
Confidence: high

File: logging/impl/src/main/java/iu/logging/internal/IuLogHandler.java [hunk 5]
```diff
@@ -355,7 +355,10 @@ public void publish(LogRecord record) {
 			return;
 
 		final var event = new IuLogEvent(record);
-		ProcessLogger.trace(() -> event.getMessage());
+		// ProcessLogger already records its own lifecycle messages. Capturing them
+		// here would feed those records back into the active trace.
+		if (!ProcessLogger.class.getName().equals(record.getLoggerName()))
+			ProcessLogger.trace(() -> event.getMessage());
```
Comment (james-anderson-iu) [inline logging/impl/src/main/java/iu/logging/internal/IuLogHandler.java]: **🟢 SUGGESTION — Process trace promotes FINE-level message content into the WARNING and ERROR log files**

This line routes the message of every published record into the active process trace regardless of level. IuLogEvent captures the whole trace whenever a record is at WARNING or above (IuLogEvent.java:116-117) and appends it to the formatted output (lines 200-203), which IuLogHandler.fileTask writes to the error and info files and consoleTask writes to System.out. The net effect is that whatever a component chose to disclose only at FINE - request bodies, header dumps, identifiers - is republished into log destinations that are otherwise level-gated, as soon as anything in the same process logs a warning. The change on this line adds a guard for ProcessLogger's own records, which is the right instinct; it does not address level promotion generally.

**Impact:** Debug-only detail reaches operators, log shippers, and subscribe() consumers who are provisioned for warning-and-above output. Exposure is capped at 80 characters per traced message by TracedMessage.append, so this is leakage of message prefixes rather than of whole payloads.
**Conditions:** Requires FINE (or any level below WARNING) to be enabled for some logger, and a WARNING or SEVERE record published within the same followed process. The observed consumer logs FINE request descriptions inside the follow boundary (esas-java-util IuTomcatRequestFilter.java:854-855) and logs WARNING for any 4xx response (lines 878-879), so both halves occur on the same request.
**Fix:** Gate the trace call on a configurable minimum level, or record the level alongside each TracedMessage and have IuLogEvent emit only messages at or above the exported record's level.
**Verify:** Add a test that logs a FINE message inside a follow boundary, then a WARNING, and asserts the FINE text does not appear in the WARNING event's getProcessLog().
Confidence: medium

File: logging/impl/src/main/java/iu/logging/internal/ProcessLogger.java [hunk 6]
```diff
@@ -405,7 +494,10 @@ public static void trace(Supplier<String> messageSupplier) {
 		if (state == null)
 			return;
 
-		state.children.offer(new TracedMessage(message, state.depth));
+		final var joinId = ACTIVE_JOIN_ID.get();
```
Comment (james-anderson-iu) [inline logging/impl/src/main/java/iu/logging/internal/ProcessLogger.java]: **🟢 SUGGESTION — follow does not clear the join ID, so a nested process inside a joined task mislabels its own messages as forked work**

join saves and restores both thread-locals (lines 430-431, 443-452), but follow saves and restores only ACTIVE_PROCESS_STATE (line 353 and 373-378). A worker thread that joins a forked process and then opens a nested follow inside the task keeps the joining task's join ID bound, so every message the nested process traces is prefixed 'N> '. The javadoc on both join (lines 411-416) and IuLogContext.join states the prefix exists 'so that work forked onto other threads can be told apart from the forking thread's own messages' - a nested process's own messages are neither, and get labelled as belonging to a join they are not part of.

**Impact:** Merged traces attribute a nested sub-process's messages to a sibling joined task, which is the exact confusion the prefix was added to prevent. Trace formatting only; no data or control-flow consequence.
**Conditions:** Requires a joined task to open its own follow boundary. No consumer does this today, but it is the natural shape for a worker that handles a discrete unit of work.
**Fix:** Save ACTIVE_JOIN_ID alongside ACTIVE_PROCESS_STATE in follow, clear it for the duration of the nested process, and restore it in the same finally.
**Verify:** Extend testForkAndJoin with a nested follow inside the joined task and assert its traced messages carry no join prefix, while messages traced after it returns carry the original one.
Confidence: high

File: logging/api/src/main/java/edu/iu/logging/IuLogContext.java [hunk 7]
```diff
@@ -111,6 +111,62 @@ static <T> T follow(IuLogContext context, String message, UnsafeSupplier<T> supp
...
+	 * </p>
+	 *
+	 * <pre>
+	 * final var fork = IuLogContext.fork();
+	 * executor.submit(() -&gt; IuLogContext.join(fork, this::handle));
+	 * </pre>
+	 *
+	 * @return opaque handle on the process being followed; null if not following
+	 */
+	static Object fork() {
```
Comment (james-anderson-iu) [inline logging/api/src/main/java/edu/iu/logging/IuLogContext.java]: **🟢 SUGGESTION — fork hands application code an implementation-module class as an untyped handle, with no documented behaviour for a stale or foreign one**

The handle is a ProcessState instance from the class the implementation module's ModularClassLoader loaded. join recovers it with an unguarded cast performed before the try block (ProcessLogger.java:429), so two things follow that the javadoc does not mention. First, any Object that did not come from fork - including a handle forked before IuLoggingBootstrap replaced the implementation module, which is then a different ProcessState class - produces a ClassCastException and the task is never run at all. Second, while an application holds a handle it pins the implementation module's class loader, which is what IuLoggingBootstrap.destroy exists to release (IuLoggingBootstrap.java:167-171).

**Impact:** A queued task silently does not run and surfaces a ClassCastException with no indication that a stale handle caused it; a handle retained across a hot replace keeps the old logging module's class loader reachable.
**Conditions:** Requires either a caller passing a handle that did not come from fork, or a logging hot replace (new IuLoggingBootstrap(true)) while a forked handle is outstanding. Neither approved consumer calls fork or join today, so this is unexercised as shipped.
**Fix:** Document on IuLogContext.fork that the handle is valid only for the lifetime of the process that produced it and must not outlive a logging reconfiguration; in ProcessLogger.join, move the cast inside the try and reject a non-ProcessState argument with an IllegalArgumentException naming the contract, so the diagnostic points at the handle rather than at a cast.
**Verify:** Add a test asserting join with an arbitrary Object fails with a message that names the fork contract, and that the task is not invoked.
Confidence: medium

File: el/src/main/java/edu/iu/util/el/El.java [hunk 8]
```diff
@@ -402,20 +400,18 @@ else if (unlessPos == -1)
...
-						final var instant = dtf.parse(((JsonString) cval).getString(), Instant::from);
+						final var date = IuJsonAdapter.of(Date.class).fromJson(cval);
 						SimpleDateFormat df = DATE_FMT.get();
 						df.applyPattern(expression.substring(1));
-						evalContext.setResult(Json.createValue(df.format(new Date(instant.toEpochMilli()))));
-					} catch (DateTimeParseException e) {
-						// ignore
-						// will return unformatted value
+						evalContext.setResult(Json.createValue(df.format(date)));
+					} catch (Exception e) {
```
Comment (james-anderson-iu) [inline el/src/main/java/edu/iu/util/el/El.java]: **🟢 SUGGESTION — El's '#' operator now swallows date-format failures to FINE while the number branch still throws for the same class of mistake**

Both halves of case '#' apply a caller-supplied pattern. On the JsonNumber branch, DecimalFormat.applyPattern (line 400) throws IllegalArgumentException for a bad pattern and that propagates out of eval. On the JsonString branch the identical mistake at line 408 is caught by this new handler, logged at FINE, and the value passes through unformatted - as ElTest lines 511-516 pin, $.string#MM\/dd\/yyyy renders 'foo'. Two opposite outcomes for one operator, and the class javadoc at lines 93-94 documents neither: it says '#' formats numbers and ISO instant strings, not that it may quietly decline to. The catch is also wider than the failure it is written for - it takes anything thrown by IuJsonAdapter, by setResult, and by Json.createValue.

**Impact:** A template author who mistypes a date pattern gets a rendered page showing a raw ISO timestamp with only a FINE log line, whereas the same typo on a numeric pattern fails loudly. Wrong output with no signal at the default log level.
**Conditions:** Applies whenever '#' is applied to a JsonString, which the tests show is a supported and exercised case.
**Fix:** Narrow the catch to the parse/format failure it is meant to absorb (ParseException / DateTimeParseException), let an invalid pattern throw as it does on the numeric branch, and document the pass-through in the class javadoc's '#' bullet so the behaviour is part of the contract rather than an artefact.
**Verify:** Add an ElTest case asserting an invalid pattern applied to a date string fails the same way an invalid pattern applied to a number does, and that only an unparseable value passes through.
Confidence: high

File: base/src/main/java/edu/iu/IuRuntimeEnvironment.java [hunk 9]
```diff
@@ -29,96 +29,195 @@
...
+	 * @param <T>  value type
+	 * @param name property name
+	 * @return system property value if set, environment variable if not set
+	 * @param textToValueFunction converts a non-null property value to the target
+	 *                            type; not applied to null
+	 * @throws NullPointerException if the system property is blank or both are
+	 *                              missing
+	 */
+	public static <T> T env(String name, Function<String, T> textToValueFunction) {
+		return Objects.requireNonNull(envOptional(name, textToValueFunction),
```
Comment (james-anderson-iu) [inline base/src/main/java/edu/iu/IuRuntimeEnvironment.java]: **🟢 SUGGESTION — env(name, textToValueFunction) reports a present property as missing when the conversion function returns null**

envOptional applies the function to any non-null, non-blank value and returns whatever it produces, so a function that maps an unrecognised value to null is indistinguishable here from an absent property. The message then states the opposite of what happened - the property is set, and its value was rejected. The message is constructed unconditionally on every call as well, so the concatenation and the second envName round trip run even on the success path.

**Impact:** An operator debugging a typo'd enum or lookup key is told to set a property that is already set, and no diagnostic names the value that failed to convert. Nothing in the new tests exercises a conversion function other than identity, so this path has never been run.
**Conditions:** Requires a textToValueFunction that can return null for a non-blank input, passed to env(name, fn) rather than envOptional. No such call site exists in iu-java-util or in any of the five approved consumer checkouts.
**Fix:** Resolve the raw value first, fail with a distinct message when it is absent or blank, and fail with a message naming the rejected value when conversion yields null. Pass a Supplier to requireNonNull so the message is not built on the success path.
**Verify:** Add tests for env(name, fn) and envOptional(name, fn) with a non-identity function, covering both a converting and a null-returning result, and assert the message distinguishes the two.
**Verified** against env(name, fn) is Objects.requireNonNull(envOptional(name, fn), "Missing system property ... or environment variable ..."), so a function returning null for a set property produces a message stating the property is absent, the only env(name, fn) call site anywhere is IuHttp.java:90, whose function ends in Collectors.toUnmodifiableList() and therefore can throw but can never return null, every other call site uses envOptional, where returning null is the documented absent result and carries no message, iu/client/Vault.java:169 forwards a caller-supplied textToValueFunction, but to envOptional rather than env, so an externally supplied null-returning function still cannot reach the misleading message, esas-sisjee, esas-thirdparty-auth and ess-dev-environment contain no env or envOptional call taking a conversion function.
Confidence: high

File: logging/impl/src/test/java/iu/logging/internal/ProcessLoggerTest.java [hunk 10]
```diff
@@ -109,6 +138,20 @@ else if (sb.charAt(i) < ' ')
...
+	/**
+	 * Runs a task on a dedicated thread, waits for it to complete, and reports
+	 * assertion failures back to the calling thread.
+	 */
+	private static void runInWorker(Runnable task) throws Throwable {
+		final var error = new Throwable[1];
+		final var worker = new Thread(task);
+		worker.setUncaughtExceptionHandler((t, e) -> error[0] = e);
+		worker.start();
+		worker.join();
```
Comment (james-anderson-iu) [inline logging/impl/src/test/java/iu/logging/internal/ProcessLoggerTest.java]: **🟢 SUGGESTION — runInWorker joins without a timeout, so the deadlock these tests exist to catch hangs the build instead of failing it**

This helper exists to drive the new cross-thread fork/join paths, which are exactly the code that could block a worker forever if a future change puts a lock or a wait behind the thread-local handoff. An unbounded Thread.join turns that failure into a hung CI job with no test name attached, rather than a failing assertion pointing at the case. ElTest already uses the right idiom for the same class of risk - assertTimeoutPreemptively at line 172, in a test explicitly named testInvalidAmpersandDoesNotHang.

**Impact:** A regression in the code this file was written to protect produces a build that never finishes instead of a report naming the test.
**Conditions:** Applies as written; no runtime condition.
**Fix:** Use worker.join(timeoutMillis) and fail with a message naming the worker if it is still alive, or wrap each fork/join test body in assertTimeoutPreemptively as ElTest does.
**Verify:** Confirm the tests still pass, and that inserting an artificial block in the joined task produces a named failure within the timeout rather than a hang.
Confidence: high

**Review by james-anderson-iu — APPROVED** [review]
